File: lib/market-data-service.ts
/**
 * Market Data Service
 * Frontend service for integrating with Python market data APIs
 */

export interface StockData {
  symbol: string
  name: string
  price: number
  change: number
  change_percent: number
  volume: number
  market_cap?: number
  pe_ratio?: number
  last_updated?: string
}

export interface CryptoData {
  symbol: string
  name: string
  price: number
  change_24h: number
  change_percent_24h: number
  market_cap: number
  volume_24h: number
  last_updated?: string
}

export interface CurrencyRate {
  from_currency: string
  to_currency: string
  rate: number
  last_updated?: string
}

export interface NewsItem {
  title: string
  summary: string
  url: string
  source: string
  published_at: string
  sentiment?: string
}

export interface MarketOverview {
  stocks: StockData[]
  crypto: CryptoData[]
  currencies: CurrencyRate[]
  news: NewsItem[]
  last_updated: string
}

export interface APIStatus {
  status: string
  message: string
  timestamp: string
}

export class MarketDataService {
  private static instance: MarketDataService
  private baseUrl: string
  private timeout: number

  constructor() {
    this.baseUrl = process.env.NEXT_PUBLIC_ML_SERVER_URL || 'http://localhost:8001'
    this.timeout = 15000 // 15 seconds for market data requests
  }

  static getInstance(): MarketDataService {
    if (!MarketDataService.instance) {
      MarketDataService.instance = new MarketDataService()
    }
    return MarketDataService.instance
  }

  /**
   * Check if the market data service is available
   */
  async isServiceAvailable(): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/api/market/health`, {
        method: 'GET',
        signal: AbortSignal.timeout(5000) // 5 second timeout for health check
      })
      return response.ok
    } catch (error) {
      console.error('Market data service health check failed:', error)
      return false
    }
  }

  /**
   * Get real-time stock data for a specific symbol
   */
  async getStockData(symbol: string): Promise<StockData | null> {
    try {
      console.log(`Fetching stock data for ${symbol}`)

      const response = await fetch(`${this.baseUrl}/api/market/stocks/${symbol.toUpperCase()}`, {
        method: 'GET',
        signal: AbortSignal.timeout(this.timeout)
      })

      if (!response.ok) {
        if (response.status === 404) {
          throw new Error(`Stock symbol ${symbol} not found`)
        }
        throw new Error(`Failed to fetch stock data: ${response.status}`)
      }

      const stockData = await response.json()
      console.log(`Stock data for ${symbol}:`, stockData)
      
      return stockData

    } catch (error) {
      console.error(`Error fetching stock data for ${symbol}:`, error)
      return null
    }
  }

  /**
   * Get stock data for multiple symbols
   */
  async getMultipleStocks(symbols?: string[]): Promise<StockData[]> {
    try {
      console.log(`Fetching multiple stocks:`, symbols)

      const url = new URL(`${this.baseUrl}/api/market/stocks`)
      if (symbols && symbols.length > 0) {
        url.searchParams.set('symbols', symbols.join(','))
      }

      const response = await fetch(url.toString(), {
        method: 'GET',
        signal: AbortSignal.timeout(this.timeout)
      })

      if (!response.ok) {
        throw new Error(`Failed to fetch stocks data: ${response.status}`)
      }

      const stocksData = await response.json()
      console.log(`Fetched ${stocksData.length} stocks`)
      
      return stocksData

    } catch (error) {
      console.error('Error fetching multiple stocks:', error)
      return []
    }
  }

  /**
   * Get cryptocurrency data
   */
  async getCryptoData(coins?: string[]): Promise<CryptoData[]> {
    try {
      console.log(`Fetching crypto data:`, coins)

      const url = new URL(`${this.baseUrl}/api/market/crypto`)
      if (coins && coins.length > 0) {
        url.searchParams.set('coins', coins.join(','))
      }

      const response = await fetch(url.toString(), {
        method: 'GET',
        signal: AbortSignal.timeout(this.timeout)
      })

      if (!response.ok) {
        throw new Error(`Failed to fetch crypto data: ${response.status}`)
      }

      const cryptoData = await response.json()
      console.log(`Fetched ${cryptoData.length} cryptocurrencies`)
      
      return cryptoData

    } catch (error) {
      console.error('Error fetching crypto data:', error)
      return []
    }
  }

  /**
   * Get currency exchange rates
   */
  async getCurrencyRates(baseCurrency: string = 'USD'): Promise<CurrencyRate[]> {
    try {
      console.log(`Fetching currency rates for base ${baseCurrency}`)

      const url = new URL(`${this.baseUrl}/api/market/currencies`)
      url.searchParams.set('base', baseCurrency.toUpperCase())

      const response = await fetch(url.toString(), {
        method: 'GET',
        signal: AbortSignal.timeout(this.timeout)
      })

      if (!response.ok) {
        throw new Error(`Failed to fetch currency rates: ${response.status}`)
      }

      const currencyData = await response.json()
      console.log(`Fetched ${currencyData.length} currency rates`)
      
      return currencyData

    } catch (error) {
      console.error('Error fetching currency rates:', error)
      return []
    }
  }

  /**
   * Get financial news
   */
  async getFinancialNews(category: string = 'business', count: number = 10): Promise<NewsItem[]> {
    try {
      console.log(`Fetching financial news: category=${category}, count=${count}`)

      const url = new URL(`${this.baseUrl}/api/market/news`)
      url.searchParams.set('category', category)
      url.searchParams.set('count', count.toString())

      const response = await fetch(url.toString(), {
        method: 'GET',
        signal: AbortSignal.timeout(this.timeout)
      })

      if (!response.ok) {
        throw new Error(`Failed to fetch financial news: ${response.status}`)
      }

      const newsData = await response.json()
      console.log(`Fetched ${newsData.length} news items`)
      
      return newsData

    } catch (error) {
      console.error('Error fetching financial news:', error)
      return []
    }
  }

  /**
   * Get comprehensive market overview
   */
  async getMarketOverview(): Promise<MarketOverview | null> {
    try {
      console.log('Fetching comprehensive market overview')

      const response = await fetch(`${this.baseUrl}/api/market/overview`, {
        method: 'GET',
        signal: AbortSignal.timeout(this.timeout)
      })

      if (!response.ok) {
        throw new Error(`Failed to fetch market overview: ${response.status}`)
      }

      const overviewData = await response.json()
      console.log('Market overview fetched:', {
        stocks: overviewData.stocks.length,
        crypto: overviewData.crypto.length,
        currencies: overviewData.currencies.length,
        news: overviewData.news.length
      })
      
      return overviewData

    } catch (error) {
      console.error('Error fetching market overview:', error)
      return null
    }
  }

  /**
   * Format currency values
   */
  formatCurrency(amount: number, currency: string = 'USD'): string {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: currency,
      minimumFractionDigits: 2,
      maximumFractionDigits: 2
    }).format(amount)
  }

  /**
   * Format percentage values
   */
  formatPercentage(value: number): string {
    const sign = value >= 0 ? '+' : ''
    return `${sign}${value.toFixed(2)}%`
  }

  /**
   * Get color for percentage changes
   */
  getChangeColor(change: number): string {
    if (change > 0) return '#10b981' // green
    if (change < 0) return '#ef4444' // red
    return '#6b7280' // gray
  }

  /**
   * Get popular stock symbols
   */
  getPopularStocks(): string[] {
    return ['AAPL', 'GOOGL', 'MSFT', 'AMZN', 'TSLA', 'META', 'NVDA', 'NFLX', 'UBER', 'SPOT']
  }

  /**
   * Get popular cryptocurrency symbols
   */
  getPopularCrypto(): string[] {
    return ['bitcoin', 'ethereum', 'binancecoin', 'cardano', 'solana', 'polkadot', 'dogecoin', 'avalanche-2']
  }

  /**
   * Get major currency codes
   */
  getMajorCurrencies(): string[] {
    return ['USD', 'EUR', 'GBP', 'JPY', 'CAD', 'AUD', 'CHF', 'CNY', 'INR']
  }

  /**
   * Format news publication date
   */
  formatNewsDate(dateString: string): string {
    const date = new Date(dateString)
    const now = new Date()
    const diffMs = now.getTime() - date.getTime()
    const diffHours = Math.floor(diffMs / (1000 * 60 * 60))
    const diffDays = Math.floor(diffHours / 24)

    if (diffHours < 1) {
      return 'Just now'
    } else if (diffHours < 24) {
      return `${diffHours}h ago`
    } else if (diffDays < 7) {
      return `${diffDays}d ago`
    } else {
      return date.toLocaleDateString()
    }
  }

  /**
   * Get sentiment color for news
   */
  getSentimentColor(sentiment?: string): string {
    switch (sentiment?.toLowerCase()) {
      case 'positive':
        return '#10b981' // green
      case 'negative':
        return '#ef4444' // red
      case 'neutral':
      default:
        return '#6b7280' // gray
    }
  }
}

// Export singleton instance
export const marketDataService = MarketDataService.getInstance()

File: server/services/market_data_service.py
"""
Market Data Service
Integrates with multiple APIs to provide real-time financial market data
Supports stocks, cryptocurrencies, currencies, and financial news
"""

import yfinance as yf
import aiohttp
import asyncio
import logging
from typing import Dict, List, Optional, Any
from datetime import datetime, timedelta
import json
import os
from dataclasses import dataclass

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

@dataclass
class StockData:
    """Stock data structure"""
    symbol: str
    name: str
    price: float
    change: float
    change_percent: float
    volume: int
    market_cap: Optional[float] = None
    pe_ratio: Optional[float] = None
    last_updated: Optional[str] = None

@dataclass
class CryptoData:
    """Cryptocurrency data structure"""
    symbol: str
    name: str
    price: float
    change_24h: float
    change_percent_24h: float
    market_cap: float
    volume_24h: float
    last_updated: Optional[str] = None

@dataclass
class CurrencyRate:
    """Currency exchange rate structure"""
    from_currency: str
    to_currency: str
    rate: float
    last_updated: Optional[str] = None

@dataclass
class NewsItem:
    """Financial news item structure"""
    title: str
    summary: str
    url: str
    source: str
    published_at: str
    sentiment: Optional[str] = None

class MarketDataService:
    """
    Advanced market data service providing real-time financial information
    """
    
    def __init__(self):
        self.session = None
        
        # Popular symbols to track
        self.popular_stocks = [
            'AAPL', 'GOOGL', 'MSFT', 'AMZN', 'TSLA', 
            'META', 'NVDA', 'NFLX', 'UBER', 'SPOT'
        ]
        
        self.popular_crypto = [
            'bitcoin', 'ethereum', 'binancecoin', 'cardano', 
            'solana', 'polkadot', 'dogecoin', 'avalanche-2'
        ]
        
        # API endpoints
        self.coingecko_api = "https://api.coingecko.com/api/v3"
        self.exchange_api = "https://api.exchangerate-api.com/v4/latest"
        self.news_api = "https://newsapi.org/v2"
        
        # Cache for API responses (simple in-memory cache)
        self.cache = {}
        self.cache_duration = 300  # 5 minutes

    async def get_session(self) -> aiohttp.ClientSession:
        """Get or create HTTP session"""
        if self.session is None or self.session.closed:
            self.session = aiohttp.ClientSession()
        return self.session

    async def close_session(self):
        """Close HTTP session"""
        if self.session and not self.session.closed:
            await self.session.close()

    def is_cache_valid(self, cache_key: str) -> bool:
        """Check if cached data is still valid"""
        if cache_key not in self.cache:
            return False
        
        cached_time = self.cache[cache_key].get('timestamp', 0)
        return (datetime.now().timestamp() - cached_time) < self.cache_duration

    def get_from_cache(self, cache_key: str) -> Optional[Any]:
        """Get data from cache if valid"""
        if self.is_cache_valid(cache_key):
            return self.cache[cache_key]['data']
        return None

    def save_to_cache(self, cache_key: str, data: Any):
        """Save data to cache with timestamp"""
        self.cache[cache_key] = {
            'data': data,
            'timestamp': datetime.now().timestamp()
        }

    async def get_stock_data(self, symbol: str) -> Optional[StockData]:
        """
        Get real-time stock data using yfinance
        """
        cache_key = f"stock_{symbol}"
        cached_data = self.get_from_cache(cache_key)
        if cached_data:
            return cached_data

        try:
            logger.info(f"Fetching stock data for {symbol}")
            
            # Create ticker object
            ticker = yf.Ticker(symbol)
            
            # Get current info
            info = ticker.info
            history = ticker.history(period="2d")
            
            if history.empty:
                logger.warning(f"No history data for {symbol}")
                return None

            current_price = history['Close'].iloc[-1]
            previous_price = history['Close'].iloc[-2] if len(history) > 1 else current_price
            
            change = current_price - previous_price
            change_percent = (change / previous_price) * 100 if previous_price != 0 else 0

            stock_data = StockData(
                symbol=symbol.upper(),
                name=info.get('longName', symbol),
                price=float(current_price),
                change=float(change),
                change_percent=float(change_percent),
                volume=int(info.get('volume', 0)),
                market_cap=info.get('marketCap'),
                pe_ratio=info.get('trailingPE'),
                last_updated=datetime.now().isoformat()
            )
            
            # Cache the result
            self.save_to_cache(cache_key, stock_data)
            return stock_data
            
        except Exception as e:
            logger.error(f"Failed to fetch stock data for {symbol}: {e}")
            return None

    async def get_multiple_stocks(self, symbols: List[str] = None) -> List[StockData]:
        """
        Get stock data for multiple symbols
        """
        if symbols is None:
            symbols = self.popular_stocks

        # Use asyncio.gather to fetch all stocks concurrently
        tasks = [self.get_stock_data(symbol) for symbol in symbols]
        results = await asyncio.gather(*tasks, return_exceptions=True)
        
        # Filter out None results and exceptions
        stock_data = []
        for result in results:
            if isinstance(result, StockData):
                stock_data.append(result)
            elif isinstance(result, Exception):
                logger.error(f"Stock data fetch error: {result}")
        
        return stock_data

    async def get_crypto_data(self, coins: List[str] = None) -> List[CryptoData]:
        """
        Get cryptocurrency data using CoinGecko API
        """
        if coins is None:
            coins = self.popular_crypto

        cache_key = f"crypto_{'_'.join(coins)}"
        cached_data = self.get_from_cache(cache_key)
        if cached_data:
            return cached_data

        try:
            logger.info(f"Fetching crypto data for {len(coins)} coins")
            
            session = await self.get_session()
            coins_str = ','.join(coins)
            url = f"{self.coingecko_api}/simple/price"
            
            params = {
                'ids': coins_str,
                'vs_currencies': 'usd',
                'include_24hr_change': 'true',
                'include_market_cap': 'true',
                'include_24hr_vol': 'true'
            }
            
            async with session.get(url, params=params) as response:
                if response.status == 200:
                    data = await response.json()
                    
                    crypto_data = []
                    for coin_id in coins:
                        if coin_id in data:
                            coin_info = data[coin_id]
                            crypto_data.append(CryptoData(
                                symbol=coin_id,
                                name=coin_id.replace('-', ' ').title(),
                                price=coin_info['usd'],
                                change_24h=coin_info.get('usd_24h_change', 0),
                                change_percent_24h=coin_info.get('usd_24h_change', 0),
                                market_cap=coin_info.get('usd_market_cap', 0),
                                volume_24h=coin_info.get('usd_24h_vol', 0),
                                last_updated=datetime.now().isoformat()
                            ))
                    
                    # Cache the result
                    self.save_to_cache(cache_key, crypto_data)
                    return crypto_data
                else:
                    logger.error(f"CoinGecko API error: {response.status}")
                    
        except Exception as e:
            logger.error(f"Failed to fetch crypto data: {e}")
        
        return []

    async def get_currency_rates(self, base_currency: str = "USD") -> List[CurrencyRate]:
        """
        Get currency exchange rates
        """
        cache_key = f"currency_{base_currency}"
        cached_data = self.get_from_cache(cache_key)
        if cached_data:
            return cached_data

        try:
            logger.info(f"Fetching currency rates for base {base_currency}")
            
            session = await self.get_session()
            url = f"{self.exchange_api}/{base_currency}"
            
            async with session.get(url) as response:
                if response.status == 200:
                    data = await response.json()
                    rates = data.get('rates', {})
                    
                    # Focus on major currencies
                    major_currencies = ['EUR', 'GBP', 'JPY', 'CAD', 'AUD', 'CHF', 'CNY', 'INR']
                    
                    currency_data = []
                    for currency in major_currencies:
                        if currency in rates and currency != base_currency:
                            currency_data.append(CurrencyRate(
                                from_currency=base_currency,
                                to_currency=currency,
                                rate=rates[currency],
                                last_updated=data.get('date', datetime.now().isoformat())
                            ))
                    
                    # Cache the result
                    self.save_to_cache(cache_key, currency_data)
                    return currency_data
                else:
                    logger.error(f"Exchange rate API error: {response.status}")
                    
        except Exception as e:
            logger.error(f"Failed to fetch currency rates: {e}")
        
        return []

    async def get_financial_news(self, category: str = "business", count: int = 10) -> List[NewsItem]:
        """
        Get financial news (placeholder implementation)
        Note: Requires NewsAPI key for full implementation
        """
        cache_key = f"news_{category}_{count}"
        cached_data = self.get_from_cache(cache_key)
        if cached_data:
            return cached_data

        # For now, return dummy news data
        # In production, integrate with NewsAPI or other news sources
        dummy_news = [
            NewsItem(
                title="Stock Market Reaches New Highs",
                summary="Major indices continue their upward trajectory as investors remain optimistic.",
                url="https://example.com/news/1",
                source="Financial Times",
                published_at=datetime.now().isoformat(),
                sentiment="positive"
            ),
            NewsItem(
                title="Cryptocurrency Market Shows Volatility",
                summary="Bitcoin and major altcoins experience significant price movements.",
                url="https://example.com/news/2", 
                source="CoinDesk",
                published_at=(datetime.now() - timedelta(hours=2)).isoformat(),
                sentiment="neutral"
            ),
            NewsItem(
                title="Federal Reserve Announces Interest Rate Decision",
                summary="Central bank maintains current monetary policy stance.",
                url="https://example.com/news/3",
                source="Reuters",
                published_at=(datetime.now() - timedelta(hours=4)).isoformat(),
                sentiment="neutral"
            )
        ]
        
        # Cache the dummy data
        self.save_to_cache(cache_key, dummy_news[:count])
        return dummy_news[:count]

    async def get_market_overview(self) -> Dict[str, Any]:
        """
        Get comprehensive market overview
        """
        try:
            logger.info("Fetching comprehensive market overview")
            
            # Fetch all data concurrently
            stocks_task = self.get_multiple_stocks()
            crypto_task = self.get_crypto_data()
            currencies_task = self.get_currency_rates()
            news_task = self.get_financial_news(count=5)
            
            stocks, crypto, currencies, news = await asyncio.gather(
                stocks_task, crypto_task, currencies_task, news_task,
                return_exceptions=True
            )
            
            # Handle exceptions
            if isinstance(stocks, Exception):
                logger.error(f"Stocks fetch failed: {stocks}")
                stocks = []
            if isinstance(crypto, Exception):
                logger.error(f"Crypto fetch failed: {crypto}")
                crypto = []
            if isinstance(currencies, Exception):
                logger.error(f"Currencies fetch failed: {currencies}")
                currencies = []
            if isinstance(news, Exception):
                logger.error(f"News fetch failed: {news}")
                news = []
            
            return {
                "stocks": [stock.__dict__ if stock else {} for stock in stocks],
                "crypto": [coin.__dict__ if coin else {} for coin in crypto],
                "currencies": [rate.__dict__ if rate else {} for rate in currencies],
                "news": [item.__dict__ if item else {} for item in news],
                "last_updated": datetime.now().isoformat()
            }
            
        except Exception as e:
            logger.error(f"Failed to fetch market overview: {e}")
            return {
                "stocks": [],
                "crypto": [],
                "currencies": [],
                "news": [],
                "last_updated": datetime.now().isoformat(),
                "error": str(e)
            }

# Global instance
market_data_service = MarketDataService()